File: y15/d09/golang/main.go
package main

import (
	"flag"
	"fmt"
	"slices"
	"strings"
)

func main() {
	f := flag.String("f", "", "Input File Name")
	flag.Parse()
	lines := readLines(*f)
	solve(lines)
}

func permute[T any](arr []T, i int, result *[][]T) {
	if i == len(arr) {
		// copy the input slice
		permutation := append([]T{}, arr...)
		*result = append(*result, permutation)
		return
	}

	for j := i; j < len(arr); j++ {
		// swap
		arr[i], arr[j] = arr[j], arr[i]
		// permute subtree
		permute(arr, i+1, result)
		// restore
		arr[i], arr[j] = arr[j], arr[i]
	}
}

func solve(lines []string) {
	graph := make(map[string]map[string]int)
	for _, line := range lines {
		parts := strings.Split(line, " ")
		u, v, d := parts[0], parts[2], convertToInt(parts[len(parts)-1])
		if graph[u] == nil {
			graph[u] = make(map[string]int)
		}
		if graph[v] == nil {
			graph[v] = make(map[string]int)
		}
		graph[u][v] = d
		graph[v][u] = d
	}

	nodes := getMapKeys(graph)
	routes := make([][]string, 0)
	permute(nodes, 0, &routes)

	distances := make([]int, 0)

	for _, route := range routes {
		routeDist := 0
		for i := 0; i < len(route)-1; i++ {
			u, v := route[i], route[i+1]
			routeDist += graph[u][v]
		}
		distances = append(distances, routeDist)
	}

	part1 := slices.Min(distances)
	fmt.Println("Part 1", part1)

	part2 := slices.Max(distances)
	fmt.Println("Part 2", part2)
}

File: y15/d09/cpp/main.cpp
#include "helper.h"
#include <algorithm>
#include <iostream>
#include <map>
#include <ostream>
#include <string>
#include <utility>
#include <vector>
using namespace std;

void permute(vector<string>& arr, int i, vector<vector<string>>& result) {
  if (i == arr.size()) {
    result.push_back(arr);
    return;
  }

  for (size_t j = i; j < arr.size() ; j++) {
    // swap
    swap(arr[i], arr[j]);
    // permute subtree 
    permute(arr, i+1, result);
    // backtrack
    swap(arr[i], arr[j]);
  }
}

int main(int argc, char *argv[]) {
  if (argc < 2) {
    cout << "Provide the input file name" << endl;
    return 1;
  }

  vector<string> lines = read_lines(argv[1]);
  map<string, map<string, int>> graph;
  
  for(const auto& line: lines) {
    vector<string> parts = split_string(line, " ");
    string u = parts[0];
    string v = parts[2];
    int d = stoi(parts[parts.size() - 1]);

    graph[u][v] = d;
    graph[v][u] = d;
  }

  vector<string> nodes;

  for (const auto& [k,v]: graph) {
    nodes.push_back(k);
  }

  vector<vector<string>> routes;
  permute(nodes, 0, routes);


  vector<int> distances;
  for(const auto& route: routes) {
    int route_dist = 0;
    for (size_t i = 0; i < route.size() - 1; i++) {
      string u = route[i];
      string v = route[i + 1];
      route_dist += graph[u][v];
    }

    distances.push_back(route_dist);
  }

  int part1 = *min_element(distances.begin(), distances.end());
  cout << "Part 1: " << part1 << endl;

  int part2 = *max_element(distances.begin(), distances.end());
  cout << "Part 2: " << part2 << endl;
  return 0;
}
